@here/harp-mapview/lib/MapView.ts

```diff
@@ -2592,18 +2585,77 @@ export class MapView extends THREE.EventDispatcher {
-        const zoomLevelDistance = cameraPosZ / Math.cos(Math.min(cameraPitch, Math.PI / 3));
-        this.m_zoomLevel = MapViewUtils.calculateZoomLevelFromDistance(this, zoomLevelDistance);
-        this.m_fog.update(this, this.m_viewRanges.maximum);
 
-        const target = MapViewUtils.getWorldTargetFromCamera(this.m_camera, this.projection);
+        //FIXME: For now we keep the old behaviour when terrain is enabled (i.e. use the camera
+        //       height above terrain to deduce the target distance).
+        //       This leads to zoomlevel changes while panning. We have to find a proper solution
+        //       for terrain (e.g. raycast with the ground surfcae that is elevated by the average
+        //       elevation in the scene)
+        const terrainEnabled = this.elevationProvider !== undefined;
+
+        // Even for a tilt of 90° raycastTargetFromCamera is returning some point almost at
+        // infinity.
+        const target =
+            !terrainEnabled && cameraPitch < PITCH_LIMIT_RAD
+                ? this.raycastTargetFromCamera()
+                : null;
         if (target !== null) {
+            // We have a valid target point. Use it to compute the distance.
             this.m_targetWorldPos.copy(target);
-            this.m_targetGeoPos = this.projection.unprojectPoint(target);
+            this.m_targetGeoPos = this.projection.unprojectPoint(this.m_targetWorldPos);
             this.m_targetDistance = this.camera.position.distanceTo(target);
+        } else {
+            // We either reached the [[PITCH_LIMIT]] or we did not hit the ground surface.
+            // In this case we do the reverse, i.e. compute some fallback distance and
+            // use it to compute the tagret point by using the camera direction.
+            const cameraPosZ = this.getCameraHeightAboveTerrain(TERRAIN_ZOOM_LEVEL);
+
+            //For flat projection we fallback to the target distance at 89 degree pitch.
+            //For spherical projection we fallback to the tangent line distance
+            this.m_targetDistance =
+                this.projection.type === ProjectionType.Planar
+                    ? cameraPosZ / Math.cos(Math.min(cameraPitch, PITCH_LIMIT_RAD))
+                    : Math.sqrt(
+                          Math.pow(cameraPosZ + EarthConstants.EQUATORIAL_RADIUS, 2) -
+                              Math.pow(EarthConstants.EQUATORIAL_RADIUS, 2)
+                      );
+
+            const cameraDir = this.m_camera.getWorldDirection(cache.vector3[0]);
+            cameraDir.multiplyScalar(this.m_targetDistance);
+            this.m_targetWorldPos.copy(this.m_camera.position).add(cameraDir);
+            this.m_targetGeoPos = this.projection.unprojectPoint(this.m_targetWorldPos);
+        }
+        this.m_zoomLevel = MapViewUtils.calculateZoomLevelFromDistance(this, this.m_targetDistance);
+    }
+
+    private raycastTargetFromCamera(): THREE.Vector3 | null {
```

Comment:
I far as I see, this function is not much different then `MapViewUtils.getWorldTargetFromCamera()`, why not use this one? Otherwise please move (reneme) it there, I will need it in ClipPlaneEvaluator(s) and Atmosphere component too.

Reply:
I want to get rid of `MapViewUtils.getWorldTargetFromCamera()`. I don't see any usage outside of `MapView`. I also marked it as deprecated.